Descripcion de columnas - Sales:
- **MonthYear**: fecha
- **Time index**: id del mes
- **Country**: país de la tienda
- **StoreID**: Id de la tienda
- **City**: Nombre de la tienda según ciudad y número interno
- **Dept_ID**: ID del departamento de la tienda
- **Dept. Name**: Nombre del departamento
- **HoursOwn**: Cantidad de horas de los trabajadores internos
- **HoursLease**: Horas trabajadas por empleados externos
- **Sales units**: Cantidad de ventas
- **Turnover**: Facturación
- **Customer**: Cantidad de clientes
- **Área (m2)**: Area del departamento
- **Opening hours**: Esquema de horas de apertura

### Importación y lectura de archivos

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
path = r'C:\Users\mateo\OneDrive\Escritorio\Ciencia de datos\Proyectos\Retail Industry\salesworkload.xlsx'

In [3]:
sales = pd.read_excel(path, sheet_name= 0)
openings = pd.read_excel(path, sheet_name= 1)

In [4]:
sales.head()

Unnamed: 0  Unnamed: 1      Unnamed: 2 Unnamed: 3  Unnamed: 4 Unnamed: 5  \
0  MonthYear  Time index         Country    StoreID        City    Dept_ID   
1    10.2016           1  United Kingdom      88253  London (I)          1   
2    10.2016           1  United Kingdom      88253  London (I)          2   
3    10.2016           1  United Kingdom      88253  London (I)          3   
4    10.2016           1  United Kingdom      88253  London (I)          4   

   Unnamed: 6 Unnamed: 7  Unnamed: 8   Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0  Dept. Name   HoursOwn  HoursLease  Sales units    Turnover    Customer   
1         Dry   3184.764           0       398560     1226244         NaN   
2      Frozen   1582.941           0        82725      387810         NaN   
3       other     47.205           0       438400      654657         NaN   
4        Fish   1623.852           0       309425      499434         NaN   

  Unnamed: 12    Unnamed: 13  
0   Area (m2)  Opening hours  
1      953.04         Type A  
2      720.48         Type A  
3      966.72         Type A  
4     1053.36         Type A

Establecemos las cabeceras y aplicamos minúsculas a todo para trabajar mejor. Además, reemplazamos los espacios con guiones bajos

In [5]:
sales.columns = sales.iloc[0]
sales = sales[1:]

In [6]:
sales.columns = sales.columns.str.lower().str.replace(' ', '_')

In [7]:
sales.tail()

0    monthyear time_index country storeid        city dept_id  \
7654   06.2017          9  Sweden   29650  Gothenburg      12   
7655   06.2017          9  Sweden   29650  Gothenburg      16   
7656   06.2017          9  Sweden   29650  Gothenburg      11   
7657   06.2017          9  Sweden   29650  Gothenburg      17   
7658   06.2017          9  Sweden   29650  Gothenburg      18   

0            dept._name  hoursown hourslease sales_units  turnover customer  \
7654           Checkout  6322.323          0     3886530  14538825      NaN   
7655  Customer Services  4270.479          0         245         0      NaN   
7656           Delivery         0          0           0         0      NaN   
7657             others  2224.929          0         245         0      NaN   
7658                all   39652.2          0     3886530  15056214      NaN   

0    area_(m2) opening_hours  
7654       #NV        Type A  
7655       #NV        Type A  
7656       #NV        Type A  
7657       #NV        Type A  
7658       #NV        Type A

In [8]:
sales.dtypes

0
monthyear        object
time_index       object
country          object
storeid          object
city             object
dept_id          object
dept._name       object
hoursown         object
hourslease       object
sales_units      object
turnover         object
customer         object
area_(m2)        object
opening_hours    object
dtype: object

In [9]:
sales.describe()

0      monthyear  time_index         country  storeid        city  dept_id  \
count       7658        7650            7650     7650        7650     7650   
unique        10           9              11       50          50       17   
top      10.2016           1  United Kingdom    88253  London (I)        1   
freq         850         850             918      153         153      450   

0      dept._name  hoursown  hourslease  sales_units  turnover customer  \
count        7650  7650.000        7650         7650      7650        0   
unique         17  3505.000         223         5658      5904        0   
top           Dry    47.205           0            0         0      NaN   
freq          450   450.000        7172          157      1480      NaN   

0       area_(m2) opening_hours  
count        7650          7650  
unique       2826             3  
top             0        Type A  
freq         1600          5661

### Correción de formatos

In [10]:
sales['monthyear'].unique()

array(['10.2016', ' - - - - ', '11.2016', '12.2016', '01.2017', '02.2017',
       '03.2017', '04.2017', '05.2017', '06.2017'], dtype=object)

In [11]:
sales[sales['monthyear'] == ' - - - - ']

0     monthyear time_index country storeid city dept_id dept._name hoursown  \
851    - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
1702   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
2553   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
3404   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
4255   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
5106   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
5957   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   
6808   - - - -         NaN     NaN     NaN  NaN     NaN        NaN      NaN   

0    hourslease sales_units turnover customer area_(m2) opening_hours  
851         NaN         NaN      NaN      NaN       NaN           NaN  
1702        NaN         NaN      NaN      NaN       NaN           NaN  
2553        NaN         NaN      NaN      NaN       NaN           NaN  
3404        NaN         NaN      NaN      NaN       NaN           NaN  
4255        NaN         NaN      NaN      NaN       NaN           NaN  
5106        NaN         NaN      NaN      NaN       NaN           NaN  
5957        NaN         NaN      NaN      NaN       NaN           NaN  
6808        NaN         NaN      NaN      NaN       NaN           NaN

Tenemos un conjunto de filas vacias. Las eliminamos. Estas son esas 8 filas extras que encontrábamos en esta columna al usar describe()

In [12]:
sales = sales[sales['monthyear'] != ' - - - - ']

In [13]:
sales['monthyear'] = pd.to_datetime(sales['monthyear'], format= '%m.%Y')

In [14]:
sales['month'] = sales['monthyear'].dt.strftime('%B')
sales['year'] = sales['monthyear'].dt.year

In [15]:
sales.head()

0  monthyear time_index         country storeid        city dept_id  \
1 2016-10-01          1  United Kingdom   88253  London (I)       1   
2 2016-10-01          1  United Kingdom   88253  London (I)       2   
3 2016-10-01          1  United Kingdom   88253  London (I)       3   
4 2016-10-01          1  United Kingdom   88253  London (I)       4   
5 2016-10-01          1  United Kingdom   88253  London (I)       5   

0           dept._name  hoursown hourslease sales_units turnover customer  \
1                  Dry  3184.764          0      398560  1226244      NaN   
2               Frozen  1582.941          0       82725   387810      NaN   
3                other    47.205          0      438400   654657      NaN   
4                 Fish  1623.852          0      309425   499434      NaN   
5  Fruits & Vegetables  1759.173          0      165515   329397      NaN   

0 area_(m2) opening_hours    month  year  
1    953.04        Type A  October  2016  
2    720.48        Type A  October  2016  
3    966.72        Type A  October  2016  
4   1053.36        Type A  October  2016  
5   1053.36        Type A  October  2016

In [16]:
sales = sales.drop('monthyear', axis= 1)

El identificador de mes no nos interesa mantenerla para nuestro trabajo así que vamos a meterlo en un dataframe aparte con su correspondencia para tenerlo a mano

In [17]:
month_id = sales[['time_index', 'month', 'year']].drop_duplicates().sort_values('time_index')

In [18]:
month_id['time_index'] = pd.to_numeric(month_id['time_index'])

In [19]:
month_id

0     time_index     month  year
1              1   October  2016
852            2  November  2016
1703           3  December  2016
2554           4   January  2017
3405           5  February  2017
4256           6     March  2017
5107           7     April  2017
5958           8       May  2017
6809           9      June  2017

In [20]:
sales = sales[sales.columns[1:]]

In [21]:
sales.storeid.unique()

array([88253, 38976, 17647, 22117, 73949, 18808, 71991, 86208, 23623,
       19769, 15552, 95434, 93033, 85321, 38560, 20891, 45583, 85696,
       32949, 96857, 87703, 19000, 88994, 20166, 16927, 96493, 88750,
       78450, 94153, 64983, 77348, 78325, 83160, 12227, 94882, 34378,
       42367, 86089, 98422, 79785, 63354, 85124, 73422, 91973, 19340,
       76852, 73762, 81473, 90992, 29650], dtype=object)

In [22]:
sales['storeid'] = pd.to_numeric(sales['storeid'])

In [23]:
city= sales['city']

In [24]:
sales[['city', 'store_number_dis']] = sales['city'].str.split('(', expand=True)

In [25]:
sales['city'] = sales['city'].str.strip()
sales['store_number_dis'] = sales['store_number_dis'].str.strip().str.strip(')')

In [26]:
sales.insert(3, 'store_number', sales['store_number_dis'])

In [27]:
sales = sales.drop('store_number_dis', axis= 1)

In [28]:
sales.head()

0         country  storeid    city store_number dept_id           dept._name  \
1  United Kingdom    88253  London            I       1                  Dry   
2  United Kingdom    88253  London            I       2               Frozen   
3  United Kingdom    88253  London            I       3                other   
4  United Kingdom    88253  London            I       4                 Fish   
5  United Kingdom    88253  London            I       5  Fruits & Vegetables   

0  hoursown hourslease sales_units turnover customer area_(m2) opening_hours  \
1  3184.764          0      398560  1226244      NaN    953.04        Type A   
2  1582.941          0       82725   387810      NaN    720.48        Type A   
3    47.205          0      438400   654657      NaN    966.72        Type A   
4  1623.852          0      309425   499434      NaN   1053.36        Type A   
5  1759.173          0      165515   329397      NaN   1053.36        Type A   

0    month  year  
1  October  2016  
2  October  2016  
3  October  2016  
4  October  2016  
5  October  2016

In [29]:
sales = sales.rename(columns= {'dept._name': 'dept_name'})

In [30]:
sales['dept_id'] = pd.to_numeric(sales['dept_id'])

La columna de nombre del departamento está bien

### Horas

No podemos pasar 'hourown' a formato numérico porque tiene algunas entradas que no son números. Vamos a encontrarlos

In [31]:
sales['hoursown_fixed'] = pd.to_numeric(sales['hoursown'], errors='coerce')

In [32]:
sales[sales['hoursown_fixed'].isnull()]

0     country  storeid       city store_number  dept_id dept_name hoursown  \
2967    Spain    16927  Barcelona            I        6      Meat        ?   
5890  Belgium    73762    Antwerp         None        2    Frozen        ?   

0    hourslease sales_units turnover customer area_(m2) opening_hours  \
2967          0     1799675  5557959      NaN   8408.64        Type A   
5890          0      189930  1335069      NaN       627        Type A   

0       month  year  hoursown_fixed  
2967  January  2017             NaN  
5890    April  2017             NaN

In [33]:
sales[(sales['city'] == 'Barcelona') & (sales['dept_name'] == 'Meat') & (sales['store_number'] == 'I')]

0    country  storeid       city store_number  dept_id dept_name  hoursown  \
414    Spain    16927  Barcelona            I        6      Meat  8629.074   
1265   Spain    16927  Barcelona            I        6      Meat  9198.681   
2116   Spain    16927  Barcelona            I        6      Meat  9979.137   
2967   Spain    16927  Barcelona            I        6      Meat         ?   
3818   Spain    16927  Barcelona            I        6      Meat  8660.544   
4669   Spain    16927  Barcelona            I        6      Meat  8773.836   
5520   Spain    16927  Barcelona            I        6      Meat  8377.314   
6371   Spain    16927  Barcelona            I        6      Meat  9381.207   
7222   Spain    16927  Barcelona            I        6      Meat  8367.873   

0    hourslease sales_units turnover customer area_(m2) opening_hours  \
414           0     1944590  6615339      NaN    8310.6        Type A   
1265          0     2641215  8613465      NaN   8007.36        Type A   
2116          0     2873570  9228366      NaN   8360.76        Type A   
2967          0     1799675  5557959      NaN   8408.64        Type A   
3818          0     1589640  6174309      NaN   8039.28        Type A   
4669          0     1923505  6389361      NaN   7941.24        Type A   
5520          0     2270300  7021083      NaN   8089.44        Type A   
6371          0     2053420  6941883      NaN   8255.88        Type A   
7222          0     1908040  6813732      NaN       #NV        Type A   

0        month  year  hoursown_fixed  
414    October  2016        8629.074  
1265  November  2016        9198.681  
2116  December  2016        9979.137  
2967   January  2017             NaN  
3818  February  2017        8660.544  
4669     March  2017        8773.836  
5520     April  2017        8377.314  
6371       May  2017        9381.207  
7222      June  2017        8367.873

Como vemos para el primer caso, las horas son muy parecidas todos los meses así que vamos a poblar este valor. En este caso considero más adecuado utilizar la mediana para poblar la celda porque
las horas de la celda 2116 actuar como outlier para la media.

In [34]:
mediana1 = sales[(sales['city'] == 'Barcelona') & (sales['dept_name'] == 'Meat') & (sales['store_number'] == 'I')]['hoursown_fixed'].median()

In [35]:
sales.loc[2967, 'hoursown_fixed'] = mediana1

Veamos el caso para la entrada de bélgica

In [36]:
sales[(sales['city'] == 'Antwerp') & (sales['dept_name'] == 'Frozen')]

0     country  storeid     city store_number  dept_id dept_name  hoursown  \
784   Belgium    73762  Antwerp         None        2    Frozen  2426.337   
1635  Belgium    73762  Antwerp         None        2    Frozen  3502.611   
2486  Belgium    73762  Antwerp         None        2    Frozen   3052.59   
3337  Belgium    73762  Antwerp         None        2    Frozen  2838.594   
4188  Belgium    73762  Antwerp         None        2    Frozen  2753.625   
5039  Belgium    73762  Antwerp         None        2    Frozen  2854.329   
5890  Belgium    73762  Antwerp         None        2    Frozen         ?   
6741  Belgium    73762  Antwerp         None        2    Frozen  2772.507   
7592  Belgium    73762  Antwerp         None        2    Frozen  2398.014   

0    hourslease sales_units turnover customer area_(m2) opening_hours  \
784           0      182250  1127841      NaN    679.44        Type A   
1635          0      196295  1226718      NaN    665.76        Type A   
2486          0      294060  1949532      NaN       627        Type A   
3337          0      139440   931953      NaN       627        Type A   
4188          0      166370  1074981      NaN    640.68        Type A   
5039          0      193010  1196646      NaN    647.52        Type A   
5890          0      189930  1335069      NaN       627        Type A   
6741          0      181065  1155267      NaN    658.92        Type A   
7592          0      164835  1112340      NaN       #NV        Type A   

0        month  year  hoursown_fixed  
784    October  2016        2426.337  
1635  November  2016        3502.611  
2486  December  2016        3052.590  
3337   January  2017        2838.594  
4188  February  2017        2753.625  
5039     March  2017        2854.329  
5890     April  2017             NaN  
6741       May  2017        2772.507  
7592      June  2017        2398.014

In [37]:
mediana2 = sales[(sales['city'] == 'Antwerp') & (sales['dept_name'] == 'Frozen')]['hoursown_fixed'].median()

Vamos a tomar la mediana de nuevo

In [38]:
sales.reset_index(drop= True, inplace= True)

In [39]:
sales.loc[5883,'hoursown_fixed'] = mediana2

In [40]:
sales = sales.drop('hoursown', axis= 1)
sales.insert(6, 'hoursown', sales['hoursown_fixed'])
sales = sales.drop('hoursown_fixed', axis= 1)

In [41]:
sales.head()

0         country  storeid    city store_number  dept_id            dept_name  \
0  United Kingdom    88253  London            I        1                  Dry   
1  United Kingdom    88253  London            I        2               Frozen   
2  United Kingdom    88253  London            I        3                other   
3  United Kingdom    88253  London            I        4                 Fish   
4  United Kingdom    88253  London            I        5  Fruits & Vegetables   

0  hoursown hourslease sales_units turnover customer area_(m2) opening_hours  \
0  3184.764          0      398560  1226244      NaN    953.04        Type A   
1  1582.941          0       82725   387810      NaN    720.48        Type A   
2    47.205          0      438400   654657      NaN    966.72        Type A   
3  1623.852          0      309425   499434      NaN   1053.36        Type A   
4  1759.173          0      165515   329397      NaN   1053.36        Type A   

0    month  year  
0  October  2016  
1  October  2016  
2  October  2016  
3  October  2016  
4  October  2016

In [42]:
sales['hourslease'] = pd.to_numeric(sales['hourslease'])

In [43]:
sales.insert(8, 'totalhours', sales['hoursown'] + sales['hourslease'])

In [44]:
sales.head()

0         country  storeid    city store_number  dept_id            dept_name  \
0  United Kingdom    88253  London            I        1                  Dry   
1  United Kingdom    88253  London            I        2               Frozen   
2  United Kingdom    88253  London            I        3                other   
3  United Kingdom    88253  London            I        4                 Fish   
4  United Kingdom    88253  London            I        5  Fruits & Vegetables   

0  hoursown  hourslease  totalhours sales_units turnover customer area_(m2)  \
0  3184.764           0    3184.764      398560  1226244      NaN    953.04   
1  1582.941           0    1582.941       82725   387810      NaN    720.48   
2    47.205           0      47.205      438400   654657      NaN    966.72   
3  1623.852           0    1623.852      309425   499434      NaN   1053.36   
4  1759.173           0    1759.173      165515   329397      NaN   1053.36   

0 opening_hours    month  year  
0        Type A  October  2016  
1        Type A  October  2016  
2        Type A  October  2016  
3        Type A  October  2016  
4        Type A  October  2016

### Ventas e ingreso

In [45]:
sales['sales_units'] = pd.to_numeric(sales['sales_units'])

In [46]:
sales['turnover'] = pd.to_numeric(sales['turnover'])
sales.rename(columns= {'turnover': 'turnover_usd'}, inplace= True)

In [47]:
sales['customer'].isnull().sum()

np.int64(7650)

La columna customer está entera vacia así que vamos a prescindir de ella

In [48]:
sales = sales.drop('customer', axis= 1)

### Area

Podemos comprobar que la columna de área tiene vacías muchas filas al final del dataframe, esto es porque el área ya se especifica en las primeras filas así que para esas útlimas va a ser información repetida

In [49]:
missings = sales[sales['area_(m2)'] == '#NV']

In [50]:
missings.head()

0            country  storeid    city store_number  dept_id  \
6800  United Kingdom    88253  London            I        1   
6801  United Kingdom    88253  London            I        2   
6802  United Kingdom    88253  London            I        3   
6803  United Kingdom    88253  London            I        4   
6804  United Kingdom    88253  London            I        5   

0               dept_name  hoursown  hourslease  totalhours  sales_units  \
6800                  Dry  2715.861           0    2715.861       341220   
6801               Frozen  1025.922           0    1025.922        70670   
6802                other    47.205           0      47.205       456405   
6803                 Fish  1189.566           0    1189.566       374335   
6804  Fruits & Vegetables  1315.446           0    1315.446       199445   

0     turnover_usd area_(m2) opening_hours month  year  
6800       1443366       #NV        Type A  June  2017  
6801        395772       #NV        Type A  June  2017  
6802        634425       #NV        Type A  June  2017  
6803         60054       #NV        Type A  June  2017  
6804         42909       #NV        Type A  June  2017

In [51]:
sales['area_(m2)'] = sales.groupby('storeid')['area_(m2)'].transform('first')

In [52]:
sales['area_(m2)'] = pd.to_numeric(sales['area_(m2)'])

In [53]:
sales.head()

0         country  storeid    city store_number  dept_id            dept_name  \
0  United Kingdom    88253  London            I        1                  Dry   
1  United Kingdom    88253  London            I        2               Frozen   
2  United Kingdom    88253  London            I        3                other   
3  United Kingdom    88253  London            I        4                 Fish   
4  United Kingdom    88253  London            I        5  Fruits & Vegetables   

0  hoursown  hourslease  totalhours  sales_units  turnover_usd  area_(m2)  \
0  3184.764           0    3184.764       398560       1226244     953.04   
1  1582.941           0    1582.941        82725        387810     953.04   
2    47.205           0      47.205       438400        654657     953.04   
3  1623.852           0    1623.852       309425        499434     953.04   
4  1759.173           0    1759.173       165515        329397     953.04   

0 opening_hours    month  year  
0        Type A  October  2016  
1        Type A  October  2016  
2        Type A  October  2016  
3        Type A  October  2016  
4        Type A  October  2016

## Limpieza de openings

Descripción de columnas - openings:
- **id**: Id de la tienda
- **Store name**: id de la tienda + nombre y número interno
- **Region**: País
- **Scheme**: Horarios
- **Month-by-month**: horas totales abierto por cada mes (hasta la columna 16)

*A partir de la columna 16 vienen otra vez todas las horas por mes pero acumuladas

Atendiendo a lo anterior, este archivo solo contiene las horas totales mes por mes de cada tienda como información adicional. Por tanto, lo que nos interesa es añadir esto a nuestra tabla de sales para tener ahí toda la información recogida.

In [54]:
openings.head()

Unnamed: 0  Unnamed: 1 Unnamed: 2 Unnamed: 3      Unnamed: 4  Unnamed: 5  \
0        NaN         NaN        NaN        NaN             NaN         NaN   
1        NaN         NaN        NaN        NaN             NaN         NaN   
2        NaN         NaN        NaN        NaN  Month-by-month         NaN   
3        NaN         NaN        NaN        NaN         10.2016     11.2016   
4         id  Store name     Region     Scheme              10     11.0000   

   Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  ...  Unnamed: 19  \
0         NaN         NaN         NaN         NaN  ...          NaN   
1         NaN         NaN         NaN         NaN  ...          NaN   
2         NaN         NaN         NaN         NaN  ...          NaN   
3     12.2016      1.2017      2.2017      3.2017  ...      12.2016   
4     12.0000      1.0000      2.0000      3.0000  ...      12.0000   

   Unnamed: 20  Unnamed: 21  Unnamed: 22  Unnamed: 23  Unnamed: 24  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3       1.2017       2.2017       3.2017       4.2017       5.2017   
4       1.0000       2.0000       3.0000       4.0000       5.0000   

   Unnamed: 25 Unnamed: 26  Unnamed: 27  Unnamed: 28  
0          NaN         NaN          NaN          NaN  
1          NaN         NaN          NaN          NaN  
2          NaN         NaN          NaN          NaN  
3       6.2017      7.2017       8.2017       9.2017  
4       6.0000      7.0000       8.0000       9.0000  

[5 rows x 29 columns]

In [55]:
openings.iloc[4, 4:] = openings.iloc[3, 4:]

In [56]:
openings.columns = openings.iloc[4]

In [57]:
openings = openings.iloc[5:, :]

In [58]:
openings.reset_index(drop= True, inplace= True)

In [59]:
openings.head()

4     id        Store name          Region  Scheme 10.2016  11.2016  12.2016  \
0  88253  88253 London (I)  United Kingdom  Type A     382    367.0    350.0   
1  38976  38976 Manchester  United Kingdom  Type A     382    367.0    350.0   
2  17647   17647 Liverpool  United Kingdom  Type A     382    367.0    350.0   
3  22117  22117 Birmingham  United Kingdom  Type A     382    367.0    350.0   
4  73949   73949 Leicester  United Kingdom  Type B     342    342.0    313.0   

4  1.2017  2.2017  3.2017  ...  12.2016  1.2017  2.2017  3.2017  4.2017  \
0   382.0   352.0   380.0  ...   1099.0  1481.0  1833.0  2213.0  2565.0   
1   382.0   352.0   380.0  ...   1099.0  1481.0  1833.0  2213.0  2565.0   
2   382.0   352.0   380.0  ...   1099.0  1481.0  1833.0  2213.0  2565.0   
3   382.0   352.0   380.0  ...   1099.0  1481.0  1833.0  2213.0  2565.0   
4   342.0   314.0   341.0  ...    997.0  1339.0  1653.0  1994.0  2308.0   

4  5.2017  6.2017  7.2017  8.2017  9.2017  
0  2932.0  3267.0  3664.0  4059.0  4426.0  
1  2932.0  3267.0  3664.0  4059.0  4426.0  
2  2932.0  3267.0  3664.0  4059.0  4426.0  
3  2932.0  3267.0  3664.0  4059.0  4426.0  
4  2636.0  2935.0  3291.0  3646.0  3974.0  

[5 rows x 29 columns]

In [60]:
sales['monthyear'] = sales['month'] + ' ' + sales['year'].astype(str)

In [61]:
sales['monthyear'] = pd.to_datetime(sales['monthyear'], format= '%B %Y')

In [64]:
sales['id_monthyear'] = sales['storeid'].astype(str) + '_' + sales['monthyear'].dt.strftime('%#m.%Y')

Este será el identificador

In [94]:
openings.columns

Index([        'id', 'Store name',     'Region',     'Scheme',    '10.2016',
            11.2016,      12.2016,       1.2017,       2.2017,       3.2017,
             4.2017,       5.2017,       6.2017,       7.2017,       8.2017,
             9.2017,          nan,    '10.2016',      11.2016,      12.2016,
             1.2017,       2.2017,       3.2017,       4.2017,       5.2017,
             6.2017,       7.2017,       8.2017,       9.2017],
      dtype='object', name=4)

Vamos a quitar las columnas de horas acumuladas ya que no las necesitamos. Si fuera necesario se podría calcular de nuevo más tarde o volver a llamar.

De la misma manera vamos a quitar las columnas relativas a Julio, Agosto y Septiembre de 2017 porque en el archivo de sales no hemos recibido la información de esos meses.

In [106]:
openings_fixed = openings.iloc[:, :13]

Ahora que ya tenemos la columna monthyear para usarla como identificador, vamos a derretir openings para establecer los meses y año como filas

In [111]:
openings_melted = openings_fixed.melt(id_vars=['id', 'Store name', 'Region', 'Scheme'], var_name= 'monthyear', value_name= 'hoursopened_pm')

In [112]:
openings_melted.head()

id        Store name          Region  Scheme monthyear hoursopened_pm
0  88253  88253 London (I)  United Kingdom  Type A   10.2016            382
1  38976  38976 Manchester  United Kingdom  Type A   10.2016            382
2  17647   17647 Liverpool  United Kingdom  Type A   10.2016            382
3  22117  22117 Birmingham  United Kingdom  Type A   10.2016            382
4  73949   73949 Leicester  United Kingdom  Type B   10.2016            342

In [114]:
openings_melted['id_monthyear'] = openings_melted['id'].astype(str) + '_' + openings_melted['monthyear'].astype(str)

In [115]:
openings_melted.head()

id        Store name          Region  Scheme monthyear hoursopened_pm  \
0  88253  88253 London (I)  United Kingdom  Type A   10.2016            382   
1  38976  38976 Manchester  United Kingdom  Type A   10.2016            382   
2  17647   17647 Liverpool  United Kingdom  Type A   10.2016            382   
3  22117  22117 Birmingham  United Kingdom  Type A   10.2016            382   
4  73949   73949 Leicester  United Kingdom  Type B   10.2016            342   

    id_monthyear  
0  88253_10.2016  
1  38976_10.2016  
2  17647_10.2016  
3  22117_10.2016  
4  73949_10.2016

In [117]:
df_final = sales.merge(openings_melted[['id_monthyear', 'hoursopened_pm']], on= 'id_monthyear')

In [118]:
df_final.head()

country  storeid    city store_number  dept_id            dept_name  \
0  United Kingdom    88253  London            I        1                  Dry   
1  United Kingdom    88253  London            I        2               Frozen   
2  United Kingdom    88253  London            I        3                other   
3  United Kingdom    88253  London            I        4                 Fish   
4  United Kingdom    88253  London            I        5  Fruits & Vegetables   

   hoursown  hourslease  totalhours  sales_units  turnover_usd  area_(m2)  \
0  3184.764           0    3184.764       398560       1226244     953.04   
1  1582.941           0    1582.941        82725        387810     953.04   
2    47.205           0      47.205       438400        654657     953.04   
3  1623.852           0    1623.852       309425        499434     953.04   
4  1759.173           0    1759.173       165515        329397     953.04   

  opening_hours    month  year  monthyear   id_monthyear hoursopened_pm  
0        Type A  October  2016 2016-10-01  88253_10.2016            382  
1        Type A  October  2016 2016-10-01  88253_10.2016            382  
2        Type A  October  2016 2016-10-01  88253_10.2016            382  
3        Type A  October  2016 2016-10-01  88253_10.2016            382  
4        Type A  October  2016 2016-10-01  88253_10.2016            382

In [119]:
depts = df_final[['dept_id', 'dept_name']]

In [122]:
df_final = df_final[['storeid', 'city', 'store_number', 'country', 'dept_name', 'hoursown', 'hourslease', 'totalhours', 'sales_units', 'turnover_usd', 'area_(m2)', 'opening_hours',
                    'hoursopened_pm', 'month', 'year']]

Con esto ya estaría el conjunto de datos relativamente limpio. 'df_final' sería el dataframe final aunque hemos creado durante el desarrollo 'city', 'month_id' y 'depts' que pueden resultar útiles para ciertos análisis. 

Para aplicar en este caso modelos predictivos de data science también deberiamos hacer un poco más de trabajo para manejar las variables categóricas.

No obstante, el dataset ha quedado libre de valores nulos, missings, gaps, incoherencias e información desordenada.